File: lib/core/models/customer.dart
class Customer {
  final int? id;
  final String name;
  final String phone;
  final String? email;
  final String? address;
  final DateTime? dob;
  final DateTime? anniversary;
  final int loyaltyPoints;
  final double totalPurchaseValue;
  final DateTime createdAt;

  Customer({
    this.id,
    required this.name,
    required this.phone,
    this.email,
    this.address,
    this.dob,
    this.anniversary,
    this.loyaltyPoints = 0,
    this.totalPurchaseValue = 0.0,
    required this.createdAt,
  });

  Map<String, dynamic> toMap() => {
        'id': id,
        'name': name,
        'phone': phone,
        'email': email,
        'address': address,
        'dob': dob?.toIso8601String(),
        'anniversary': anniversary?.toIso8601String(),
        'loyalty_points': loyaltyPoints,
        'total_purchase_value': totalPurchaseValue,
        'created_at': createdAt.toIso8601String(),
      };

  factory Customer.fromMap(Map<String, dynamic> map) => Customer(
        id: map['id'],
        name: map['name'],
        phone: map['phone'],
        email: map['email'],
        address: map['address'],
        dob: map['dob'] != null ? DateTime.parse(map['dob']) : null,
        anniversary: map['anniversary'] != null ? DateTime.parse(map['anniversary']) : null,
        loyaltyPoints: map['loyalty_points'] ?? 0,
        totalPurchaseValue: (map['total_purchase_value'] ?? 0.0).toDouble(),
        createdAt: DateTime.parse(map['created_at']),
      );
}

File: backend old/models/Customer.ts
import mongoose from "mongoose";

const CustomerSchema = new mongoose.Schema({
  name: { type: String, required: true },
  email: { type: String, lowercase: true },
  phone: { type: String, required: true },
  address: { type: String },
  city: { type: String },
  state: { type: String },
  pincode: { type: String },
  pan: { type: String, uppercase: true },
  panNumber: { type: String, uppercase: true, trim: true, index: true },
  panStatus: {
    type: String,
    enum: ["PENDING", "VERIFIED", "INVALID"],
    default: "PENDING",
    index: true,
  },
  panVerifiedAt: { type: Date },
  panVerifiedBy: { type: String },
  tenantId: { type: String, default: "default-shop", index: true },
  aadhar: { type: String },
  gstin: { type: String },
  kycStatus: { type: String, enum: ["PENDING", "VERIFIED", "REJECTED"], default: "PENDING" },
  kycDocuments: [{
    type: { type: String, enum: ["PAN", "AADHAR", "PASSPORT"] },
    documentPath: String,
    uploadedAt: Date
  }],
  creditLimit: { type: Number, default: 0 },
  creditBlocked: { type: Boolean, default: false },
  advanceBalance: { type: Number, default: 0, min: 0 },
  outstandingBalance: { type: Number, default: 0, min: 0 },
  loyaltyPoints: { type: Number, default: 0 },
  totalPurchases: { type: Number, default: 0 },
  customerSince: { type: Date, default: Date.now },
  ledgerAccountId: { type: String },
  loyaltyWalletId: { type: String },
  preferredBranch: { type: mongoose.Schema.Types.ObjectId, ref: "Branch" },
  birthday: { type: Date },
  anniversary: { type: Date },
  referredBy: { type: String },
  customerTier: { type: String, enum: ["REGULAR", "NORMAL", "VIP", "BLACKLISTED"], default: "NORMAL" },
  vipSince: { type: Date },
  blacklistReason: { type: String },
  blacklistDate: { type: Date },
  blacklistedBy: { type: String },
  tierNotes: { type: String },
  tags: { type: [String], default: ["REGULAR"] },
  status: { type: String, enum: ["ACTIVE", "INACTIVE"], default: "ACTIVE" },
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now }
});

CustomerSchema.index({ phone: 1 });
CustomerSchema.index({ email: 1 });
CustomerSchema.index({ pan: 1 });
CustomerSchema.index({ tenantId: 1, panNumber: 1 }, { unique: true, sparse: true });

const Customer = (mongoose.models.Customer || mongoose.model("Customer", CustomerSchema)) as mongoose.Model<any>;
export default Customer;
